### PR TITLE
Airflow agent

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
Airflow agent allows you to seamlessly run Airflow tasks in the Flyte workflow without changing code. All the airflow tasks will be run on an airflow agent (long-running server) instead of launching a new pod to run it, which dramatically reduces overhead.

- Compile Airflow tasks to Flyte tasks
- Use Airflow sensors/operators in Flyte workflows
- Add support running Airflow tasks locally without running a cluster

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

DataprocOperator
```python
from datetime import timedelta

from airflow.utils import trigger_rule

from flytekit import task, workflow, ImageSpec
from airflow.sensors import time_sensor
from airflow.providers.google.cloud.operators.dataproc import DataprocCreateClusterOperator, DataprocDeleteClusterOperator, DataprocSubmitSparkJobOperator

airflow_plugin = "git+https://github.com/flyteorg/flytekit.git@487438ab59147879eded897674593a1eaee1c78b#subdirectory=plugins/flytekit-airflow"
image_spec = ImageSpec(base_image="pingsutw/flytekit:v1", packages=["apache-airflow", airflow_plugin], apt_packages=["git"], registry="pingsutw")


@workflow
def wf():
    create_cluster = DataprocCreateClusterOperator(
        task_id="create_dataproc_cluster",
        image_version="2.0.27-debian10",
        storage_bucket="opta-gcp-dogfood-gcp",
        master_machine_type="n1-highmem-32",
        master_disk_size=1024,
        num_workers=2,
        worker_machine_type="n1-highmem-64",
        worker_disk_size=1024,
        region="us-west1",
        cluster_name="flyte-dataproc",
        project_id="dogfood-gcp-dataplane",
    )

    run_spark = DataprocSubmitSparkJobOperator(
        job_name="spark_pi",
        task_id="run_spark",
        dataproc_jars=["file:///usr/lib/spark/examples/jars/spark-examples.jar"],
        main_class="org.apache.spark.examples.JavaWordCount",
        arguments=["gs://opta-gcp-dogfood-gcp/spark/file.txt"],
        cluster_name="flyte-dataproc",
        region="us-west1",
        project_id="dogfood-gcp-dataplane",
    )

    delete_cluster = DataprocDeleteClusterOperator(
        task_id="create_dataproc_cluster",
        project_id="dogfood-gcp-dataplane",
        cluster_name="flyte-dataproc",
        region="us-west1",
        retries=3,
        retry_delay=timedelta(minutes=5),
        email_on_failure=True,
        trigger_rule=trigger_rule.TriggerRule.ALL_DONE
    )

    create_cluster >> run_spark >> delete_cluster


if __name__ == '__main__':
    wf()
```

**bashSensor**
```python
from airflow.sensors.bash import BashSensor
from flytekit import task, workflow, ImageSpec


airflow_plugin = "git+https://github.com/flyteorg/flytekit.git@487438ab59147879eded897674593a1eaee1c78b#subdirectory=plugins/flytekit-airflow"
image_spec = ImageSpec(base_image="pingsutw/flytekit:v1", packages=["apache-airflow", airflow_plugin], apt_packages=["git"], registry="pingsutw")


@task(container_image=image_spec)
def t1():
    print("flyte")


@workflow
def wf():
    sensor = BashSensor(task_id="Sensor_succeeds", bash_command="echo hello")
    sensor >> t1()


if __name__ == '__main__':
    wf()
```

**HiveSensor**
```python
from airflow.providers.apache.hive.sensors.hive_partition import HivePartitionSensor

from flytekit import task, workflow, ImageSpec


airflow_plugin = "git+https://github.com/flyteorg/flytekit.git@487438ab59147879eded897674593a1eaee1c78b#subdirectory=plugins/flytekit-airflow"
image_spec = ImageSpec(base_image="pingsutw/flytekit:v1", packages=["apache-airflow", airflow_plugin], apt_packages=["git"], registry="pingsutw")


@task(container_image=image_spec)
def t1():
    print("flyte")


@workflow
def wf():
    sensor = HivePartitionSensor(table="flyte", schema="person")
    sensor >> t1()


if __name__ == '__main__':
    wf()
```

**FileSensor**
```python
from airflow.sensors.filesystem import FileSensor
from flytekit import task, workflow, ImageSpec


airflow_plugin = "git+https://github.com/flyteorg/flytekit.git@487438ab59147879eded897674593a1eaee1c78b#subdirectory=plugins/flytekit-airflow"
image_spec = ImageSpec(base_image="pingsutw/flytekit:v1", packages=["apache-airflow", airflow_plugin], apt_packages=["git"], registry="pingsutw")


@task(container_image=image_spec)
def t1():
    print("flyte")


@workflow
def wf():
    sensor = FileSensor(task_id="id", filepath="/tmp/1234")
    sensor >> t1()


if __name__ == '__main__':
    wf()
```

**TimeSensor**

```python
from datetime import datetime, timedelta
from pytz import UTC

from airflow.sensors.time_sensor import TimeSensor
from flytekit import task, workflow, ImageSpec

airflow_plugin = "git+https://github.com/flyteorg/flytekit.git@487438ab59147879eded897674593a1eaee1c78b#subdirectory=plugins/flytekit-airflow"
image_spec = ImageSpec(base_image="pingsutw/flytekit:v1", packages=["apache-airflow", airflow_plugin], apt_packages=["git"], registry="pingsutw")


@task(container_image=image_spec)
def t1():
    print("flyte")


@workflow
def wf():
    sensor = TimeSensor(task_id="fire_immediately", target_time=(datetime.now(tz=UTC)+timedelta(seconds=5)).time())
    sensor >> t1()


if __name__ == '__main__':
    wf()
```
**PythonSensor**
```python
from airflow.sensors.python import PythonSensor
from flytekit import task, workflow, ImageSpec

airflow_plugin = "git+https://github.com/flyteorg/flytekit.git@487438ab59147879eded897674593a1eaee1c78b#subdirectory=plugins/flytekit-airflow"
image_spec = ImageSpec(base_image="pingsutw/flytekit:v1", packages=["apache-airflow", airflow_plugin], apt_packages=["git"], registry="pingsutw")


@task(container_image=image_spec)
def t1():
    print("flyte")


def py_func():
    print("airflow python sensor")
    return True


@workflow
def wf():
    sensor = PythonSensor(task_id="fire_immediately", python_callable=py_func)
    sensor >> t1()


if __name__ == '__main__':
    wf()
```

## Tracking Issue
_NA_

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by zeryx** [comment]
Crossposting a PR review:

Airflow plugin is still required (even though we've swapped to agents), backend plugin will need to be installed
Examples run fine, and can be bundled together:
```
from datetime import datetime, timedelta
from pytz import UTC

from airflow.sensors.time_sensor import TimeSensor
from airflow.sensors.filesystem import FileSensor
from airflow.sensors.bash import BashSensor
from flytekit import workflow, ImageSpec
@workflow
def wf():
    sensor = TimeSensor(task_id="fire_immediately", target_time=(datetime.now(tz=UTC)+timedelta(seconds=5)).time())
    file_sensor = FileSensor(task_id="id", filepath="/tmp/1234")
    t3 = BashSensor(task_id="Sensor_succeeds", bash_command="exit 0")
    sensor >> file_sensor >> t3
```

This is great! Very cool, however any operator does not function. Lets add a bashOperator to the example
```
from datetime import datetime, timedelta
from textwrap import dedent

from flytekit import workflow, task, ImageSpec
# The DAG object; we'll need this to instantiate a DAG

from datetime import datetime, timedelta
from pytz import UTC

from airflow.sensors.time_sensor import TimeSensor
from airflow.sensors.filesystem import FileSensor
from airflow.sensors.bash import BashSensor
from airflow.operators.bash import BashOperator
from flytekit import workflow, ImageSpec
@workflow
def wf():
    sensor = TimeSensor(task_id="fire_immediately", target_time=(datetime.now(tz=UTC)+timedelta(seconds=5)).time())
    file_sensor = FileSensor(task_id="id", filepath="/tmp/1234")
    t3 = BashSensor(task_id="Sensor_succeeds", bash_command="exit 0")
    foo = BashOperator(task_id="foo", bash_command="echo foo")
    sensor >> file_sensor >> t3 >> foo
```
error is
```
pyflyte run workflows/example.py  wf
Failed with Unknown Exception <class 'AttributeError'> Reason: Error encountered while executing 'wf':
  'BashOperator' object has no attribute 'ref'
  ```
I have not tested remote functionality yet @pingsutw

File: Dockerfile.agent [hunk 1]
```diff
@@ -3,7 +3,19 @@ FROM python:3.9-slim-buster
...
+
+# RUN pip install --upgrade pip setuptools wheel psutil
+RUN pip install apache-airflow
+RUN pip install msgpack
+RUN pip install google-cloud-orchestration-airflow==1.9.1
+RUN pip install apache-airflow-providers-google
+RUN pip install jsonpickle
+# RUN pip install apache-airflow-providers-apache-hive
+
+RUN pip install git+https://github.com/flyteorg/flytekit.git@3fbdf47bc2b925672462346fdf0520b3b5541a04#subdirectory=plugins/flytekit-airflow
```
Comment (pingsutw) [inline Dockerfile.agent:18]: Note: update this before we merge it

**Comment by davidmirror-ops** [comment]
@pingsutw What's the status of this PR?

**Review by wild-endeavor — DISMISSED** [review]

File: .github/workflows/pythonbuild.yml [hunk 2]
```diff
@@ -163,6 +164,8 @@ jobs:
           # See: https://github.com/flyteorg/flytekit/actions/runs/4493746408/jobs/7905368664
           - python-version: 3.11
             plugin-names: "flytekit-whylogs"
+          - python-version: 3.11
+            plugin-names: "flytekit-airflow"
```
Comment (eapolinario) [inline .github/workflows/pythonbuild.yml:168]: curious, do we really need to skip this?

File: Dockerfile [hunk 3]
```diff
@@ -20,6 +20,7 @@ RUN pip install -U flytekit==$VERSION \
 
 RUN useradd -u 1000 flytekit
 RUN chown flytekit: /root
+RUN chown flytekit: /home
```
Comment (eapolinario) [inline Dockerfile:23]: can you leave a comment explaining why we're doing this?

File: plugins/flytekit-airflow/setup.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    install_requires=plugin_requires,
+    license="apache2",
+    python_requires=">=3.8",
+    classifiers=[
+        "Intended Audience :: Science/Research",
+        "Intended Audience :: Developers",
+        "License :: OSI Approved :: Apache Software License",
+        "Programming Language :: Python :: 3.8",
+        "Programming Language :: Python :: 3.9",
+        "Programming Language :: Python :: 3.10",
```
Comment (eapolinario) [inline plugins/flytekit-airflow/setup.py:34]: does it really not work with python 3.11?

**Review by eapolinario — COMMENTED** [review]
Just a few comments. Looks really close.

**Review by eapolinario — APPROVED** [review]
